sphinx/util/logging.py

```diff
@@ -8,9 +8,293 @@
     :copyright: Copyright 2007-2016 by the Sphinx team, see AUTHORS.
     :license: BSD, see LICENSE for details.
 """
+from __future__ import absolute_import
+
+import logging
+import logging.handlers
+from contextlib import contextmanager
+from collections import defaultdict
+
+from six import PY2, StringIO
+from docutils.utils import get_source_line
+
+from sphinx.errors import SphinxWarning
+from sphinx.util.console import colorize
+
+if False:
+    # For type annotation
+    from typing import Any, Generator, IO, Tuple  # NOQA
+    from docutils import nodes  # NOQA
+    from sphinx.application import Sphinx  # NOQA
+
+
+VERBOSE = 15
+DEBUG2 = 5
+
+LEVEL_NAMES = defaultdict(lambda: logging.WARNING)  # type: Dict[str, int]
+LEVEL_NAMES.update({
+    'CRITICAL': logging.CRITICAL,
+    'SEVERE': logging.CRITICAL,
+    'ERROR': logging.ERROR,
+    'WARNING': logging.WARNING,
+    'INFO': logging.INFO,
+    'VERBOSE': VERBOSE,
+    'DEBUG': logging.DEBUG,
+    'DEBUG2': DEBUG2,
+})
+
+VERBOSITY_MAP = defaultdict(lambda: 0)  # type: Dict[int, int]
+VERBOSITY_MAP.update({
+    0: logging.INFO,
+    1: VERBOSE,
+    2: logging.DEBUG,
+    3: DEBUG2,
+})
+
+COLOR_MAP = defaultdict(lambda text: text)  # type: Dict[int, unicode]
+COLOR_MAP.update({
+    logging.WARNING: 'darkred',
+    logging.DEBUG: 'darkgray',
+    DEBUG2: 'lightgray',
+})
+
+
+def getLogger(name):
+    # type: (str) -> SphinxLoggerAdapter
+    """Get logger wrapped by SphinxLoggerAdapter."""
+    return SphinxLoggerAdapter(logging.getLogger(name), {})
+
+
+def convert_serializable(records):
+    """Convert LogRecord serializable."""
+    for r in records:
+        # extract arguments to a message and clear them
+        r.msg = r.getMessage()
+        r.args = ()
+
+
+class SphinxWarningLogRecord(logging.LogRecord):
+    """Log record class supporting location"""
+    location = None  # type: Any
+
+    def getMessage(self):
```

Comment:
I try to implement this as LogRecord. But it seems the behavior is similar to Formatter.
Is it better to modify this to Formatter?